gmail.js

```diff
@@ -104,172 +105,144 @@ async function getEmailStats(user, emails) {
-  return new Promise((resolve, reject) => {
-    gmail.users.getProfile(
-      {
-        userId: "me",
-      },
-      (err, res) => {
-        if (err)
-          return reject(
-            "The API returned an error while fetching user: " + err
-          );
-        resolve(res.data);
-      }
-    );
-  });
+  try {
+    const res = await gmail.users.getProfile({
+      userId: "me",
+    });
+    return res.data;
+  } catch (err) {
+    throw "The API returned an error while fetching user: " + err;
+  }
 }
 
 /**
  * Lists the labels in the user's account.
- *
  * @param {google.auth.OAuth2} auth An authorized OAuth2 client.
  */
-function getLabels(auth) {
+async function getLabels(auth) {
   const gmail = google.gmail({ version: "v1", auth });
-  return new Promise((resolve, reject) => {
-    gmail.users.labels.list(
-      {
-        userId: "me",
-      },
-      (err, res) => {
-        if (err) return reject("The API returned an error " + err);
-        const labels = res.data.labels;
-        if (labels) {
-          resolve(labels);
-        } else {
-          resolve([]);
-        }
-      }
-    );
-  });
+  try {
+    const res = gmail.users.labels.list({
+      userId: "me",
+    });
+    return res.data.labels;
+  } catch (err) {
+    throw "The API returned an error: " + err;
+  }
 }
 
 /**
  * Fetches received messages in the user's account.
- *
  * @param {google.auth.OAuth2} auth An authorized OAuth2 client.
  */
-function getEmailKeys(auth, labels, hourAgo) {
+async function getEmailKeys(auth, labels, hourAgo) {
   const gmail = google.gmail({ version: "v1", auth });
-  return new Promise((resolve, reject) => {
-    gmail.users.messages.list(
-      {
-        userId: "me",
-        q: `after:${hourAgo}`,
-        labelIds: labels,
-      },
-      (err, res) => {
-        if (err) return reject("The API returned an error " + err);
-        const messages = res.data.messages;
-        if (messages) {
-          resolve(messages);
-        } else {
-          resolve([]);
-        }
-      }
-    );
-  });
+  try {
+    const res = await gmail.users.messages.list({
+      userId: "me",
+      q: `after:${hourAgo}`,
+      labelIds: labels,
+    });
+    const msgs = res.data.messages;
+    if (msgs) {
+      return msgs;
+    } else {
+      return [];
+    }
+  } catch (err) {
+    throw "The API returned an error: " + err;
+  }
 }
 
 /**
  * Process raw emails and returning parsed emails.
- *
  * @param {google.auth.OAuth2} auth An authorized OAuth2 client.
  * @param {gmail_v1.Schema$Message} rawEmails A list of raw emails retrieved from Gmail API client.
  */
-function getParsedEmails(auth, rawEmails) {
+async function getParsedEmails(auth, rawEmails) {
   const emails = [];
-  return new Promise(async (resolve) => {
-    if (rawEmails.length) {
-      for (const rawEmail of rawEmails) {
-        const email = await parseEmail(auth, rawEmail);
-        emails.push(email);
-      }
+  if (rawEmails.length) {
```

Comment:
Checkout the package called bluebird and `promise.map`. It'll let you rewrite this as

```
const emails = await Promise.map(rawEmails, email => parseEmail(auth, email))
```

Reply:
Wow, that's way cleaner. Thanks a ton.